RetroPlayer: Modernize OpenGL rendering

VelocityRa opened the pull request:
## Description
<!--- Provide a general summary of your change in the Pull Request title above -->
<!--- Describe your change in detail here -->
* Use of more modern OpenGL practices by using [Vertex Array Objects](https://www.khronos.org/opengl/wiki/Vertex_Specification#Vertex_Array_Object).
* An optimization to not generate Vertex Buffer Objects every frame, since that's unnecessary and inadvisable.
This was also done for the GLES renderer, not possible for the former change because it's not supported.

## Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here -->
* It's more modern
* It's faster

## How Has This Been Tested?
<!--- Please describe in detail how you tested your change -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc -->
**GL**: Tried 'game.libretro.2048' on RetroPlayer, it renders as it did before.
**GLES**: Untested, but the code changed is identical to the GL commit (which I tested by itself on GL and worked).

## Screenshots (if appropriate):
Not necessary, nothing visual changed.

## Types of change
<!--- What type of change does your code introduce? Put an `x` in all the boxes that apply like this: [X] -->
- [ ] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [x] **Improvement** (non-breaking change which improves existing functionality)
- [ ] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **None of the above** (please explain below)

## Checklist:
<!--- Go over all the following points, and put an `X` in all the boxes that apply like this: [X] -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [x] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [ ] I have added tests to cover my change
- [ ] All new and existing tests passed

VelocityRa committed:
RetroPlayer/renderer/GL: Don't generate VBOs every frame

VelocityRa committed:
RetroPlayer/renderer/GL: Use VAOs

VelocityRa committed:
RetroPlayer/renderer/GLES: Don't generate VBOs every frame

lrusak committed:
RetroPlayer/renderer/GBM: Don't generate VBOs every frame

VelocityRa commented:
jenkins build this please

MartijnKaijser commented:
God for merge @lrusak @garbear ?

garbear commented on xbmc/cores/RetroPlayer/rendering/VideoRenderers/RPRendererOpenGL.cpp:
```diff
@@ -306,20 +303,25 @@ void CRPRendererOpenGL::Render(uint8_t alpha)
   m_context.DisableGUIShader();
 }
+
+void CRPRendererOpenGL::Deinitialize()
```
It was a big red flag that this function overrided but didn't call superclass Deinitialize(). I see that parent implementation is empty, but you shouldn't assume it will stay that way.

garbear commented on xbmc/cores/RetroPlayer/rendering/VideoRenderers/RPRendererOpenGL.h:
```diff
@@ -44,6 +44,18 @@ namespace RETRO
   protected:
+    struct PackedVertex
+    {
+      float x, y, z;
```
I prefer vars on separate lines, but x y z together is OK.

garbear commented on xbmc/cores/RetroPlayer/rendering/VideoRenderers/RPRendererOpenGLES.cpp:
```diff
@@ -320,9 +315,14 @@ void CRPRendererOpenGLES::Render(uint8_t alpha)
   m_context.DisableGUIShader();
 }
+
+void CRPRendererOpenGLES::Deinitialize()
```
Can you add a call to parent deinitialize here too?

garbear commented:
Disregard my comments about Deinitiaize(). CRPBaseRenderer::Deinitialize() is never called and should be removed.

VelocityRa committed:
RetroPlayer/rendering: Remove Deinitialize() from RPBaseRenderer

It was easy to mistakenly call many times from subclasses, this is simpler and more idiomatic.

garbear reviewed (approved):
thanks for the last commit. looks good

VelocityRa commented:
@FernetMenta This PR was open for a month, not sure why is this being brought up now...

What do you mean by "after game was loaded and still in GUI"? How to reproduce?

FernetMenta commented:
> not sure why this is this being brought up now...

well, I tried games for the first time and fell into this trap.

> How to reproduce?

Load a game with Internet Archive Game launcher. Every time I load a game, I get a black screen for the reason mentioned. Then I hit key <back> to enter game mode.

fritsch commented:
I tried the same. NES black screen. SNES9X works as it should.

lrusak commented:
@fritsch with which emulator? I just tested with bnes and nestopia and both work for me (though bnes has some wrong colors)

VelocityRa commented:
Yeah I can't repro either. Tried all 4 available NES emulators.

fritsch commented:
I used nestopia. Window-Manager: openbox. X11 as platform.

garbear commented:
Is this the same black screen problem? https://github.com/kodi-game/game.libretro.sameboy/issues/3
